src/networks/invertible_network_unet.jl

```diff
@@ -0,0 +1,133 @@
+# Invertible network layer from Putzky and Welling (2019): https://arxiv.org/abs/1911.10914
+# Author: Philipp Witte, pwitte3@gatech.edu
+# Date: January 2020
+
+export NetworkUNET, NetworkUNET3D
+
+"""
+    L = NetworkUNET(n_in, n_hidden, maxiter, Ψ; k1=4, k2=3, p1=0, p2=1, s1=4, s2=1, ndims=2) (2D)
+
+    L = NetworkUNET3D(n_in, n_hidden, maxiter, Ψ; k1=4, k2=3, p1=0, p2=1, s1=4, s2=1) (3D)
+
+ Create an invertibel recurrent inference machine (i-RIM) consisting of an unrooled loop
+ for a given number of iterations.
+
+ *Input*: 
+ 
+ - 'n_in': number of input channels
+
+ - `n_hidden`: number of hidden units in residual blocks
+
+ - `maxiter`: number unrolled loop iterations
+
+ - `Ψ`: link function
+
+ - `k1`, `k2`: stencil sizes for convolutions in the residual blocks. The first convolution 
+   uses a stencil of size and stride `k1`, thereby downsampling the input. The second 
+   convolutions uses a stencil of size `k2`. The last layer uses a stencil of size and stride `k1`,
+   but performs the transpose operation of the first convolution, thus upsampling the output to 
+   the original input size.
+
+ - `p1`, `p2`: padding for the first and third convolution (`p1`) and the second convolution (`p2`) in
+   residual block
+
+ - `s1`, `s2`: stride for the first and third convolution (`s1`) and the second convolution (`s2`) in
+   residual block
+
+ - `ndims` : number of dimensions
+
+ *Output*:
+ 
+ - `L`: invertible i-RIM network.
+
+ *Usage:*
+
+ - Forward mode: `η_out, s_out = L.forward(η_in, s_in, d, A)`
+
+ - Inverse mode: `η_in, s_in = L.inverse(η_out, s_out, d, A)`
+
+ - Backward mode: `Δη_in, Δs_in, η_in, s_in = L.backward(Δη_out, Δs_out, η_out, s_out, d, A)`
+
+ *Trainable parameters:*
+
+ - None in `L` itself
+
+ - Trainable parameters in the invertible coupling layers `L.L[i]`, and actnorm layers
+   `L.AN[i]`, where `i` ranges from `1` to the number of loop iterations.
+
+ See also: [`CouplingLayerIRIM`](@ref), [`ResidualBlock`](@ref), [`get_params`](@ref), [`clear_grad!`](@ref)
+"""
+struct NetworkUNET <: InvertibleNetwork
+    L::CouplingLayerIRIM
+    AN::ActNorm
+    n_mem::Int64
+end
+
+@Flux.functor NetworkUNET
+
+# 2D Constructor
+function NetworkUNET(n_in::Int64, n_hiddens::Array{Int64,1}, ds::Array{Int64,1}; k1=4, k2=3, p1=0, p2=1, s1=4, s2=1, ndims=2)
```

Comment:
The order of operations do not seem the same as [https://github.com/pputzky/invertible_rim/blob/91c6659475b80d6f0a451448a0b6156fcfbecce4/irim/core/invertible_unet.py#L26](https://github.com/pputzky/invertible_rim/blob/91c6659475b80d6f0a451448a0b6156fcfbecce4/irim/core/invertible_unet.py#L26). E.g., `Conv1x1` is happening at the end in the original source code.